### ISSUE TITLE
[VMAccessForLinux-1.4.2.0]"Reset configuration only" in ARM mode results in log error

### ISSUE THREAD

**Issue by yuxisun1217** [Azure/azure-linux-extensions#235]
Description of problem:
"Reset configuration only" in ARM mode results in log error.
According to the feedback of Hans Krijger, It looks like _set_user_account_pub_key is called after _reset_sshd_config succeeds, resulting in the error.
https://github.com/Azure/azure-linux-extensions/blob/master/VMAccess/vmaccess.py#L87-L94 

Version-Release number of selected component (if applicable):
VMAccessForLinux-1.4.2.0

RHEL Version: RHEL-7.3 internal build
Python Version: 2.7.5

How reproducible:
100%

Steps to Reproduce:
1. Prepare a RHEL7.3 VM on Azure East US location(ARM mode) with WALA-2.1.6 installed.
2. Select the VM,go to "Reset password" and then "Reset configuration only", click "OK"

Actual results:
The file has been backuped and create a new file. But there is log error in log file:

/var/log/azure/Microsoft.OSTCExtensions.VMAccessForLinux/1.4.2.0/extension.log
2016/09/01 13:22:32 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.2.0] sequence number is 0
2016/09/01 13:22:32 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.2.0] setting file path is/var/lib/waagent/Microsoft.OSTCExtensions.VMAccessForLinux-1.4.2.0/config/0.settings
2016/09/01 13:22:32 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.2.0] JSON config: {"runtimeSettings": [{"handlerSettings": {"protectedSettings": "MIIB8AYJKoZIhvcNAQcDoIIB4TCCAd0CAQAxggFpMIIBZQIBADBNMDkxNzA1BgoJkiaJk/IsZAEZFidXaW5kb3dzIEF6dXJlIENSUCBDZXJ0aWZpY2F0ZSBHZW5lcmF0b3ICEGrDCpgcV9m/RyEPNDQi8XowDQYJKoZIhvcNAQEBBQAEggEARY+RtVJ0mILISBmAR3pcKD42n5yBoIMyaxB48QzmYFtcAVapWvs0LgLSqjbfAL4TgoYeyBXctVAl37Y1WyZnhvdXgxsM86k+G+cU/GGzwqAPcbSWgdDpeOs4z5TXqa0KxpODwGY9JziQL3muL7F/9NZjUGNnxdTevjMwOxo6FfyESJ/4WDjalyXQanmDiPosw0M52uClNHXEAZHwb9TFl4TzW6AGlZXoKxI/57AEaPsxKlu2CBp/BHGF+TvOsD61j/ooKCZaBfO4HxeLbttg0lX6fdTedwsI3BzuVOA3Zb71jacjyrEe7T8HxkdsdnHHhoLnOeiGcBFypUqpnNIL3zBrBgkqhkiG9w0BBwEwFAYIKoZIhvcNAwcECI8qWYNWbPJLgEiIEn7oBrOG4HXusfxDXALruhEcVBwzdzUUoQVzY8CdqZSLuQWF+6sD0qkIWl0r/Oy1lqAdDXhKtZGGbjQDHwE1i0VLSNS+LNk=", "publicSettings": {}, "protectedSettingsCertThumbprint": "DFEE761B6AF1547E433C6999FCDE5C4D8371A5F0"}}]}
2016/09/01 13:22:32 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.2.0] Config decoded correctly.
2016/09/01 13:22:33 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.2.0] Succeeded in check and open ssh port.
2016/09/01 13:22:33 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.2.0] set most recent sequence number to 0
2016/09/01 13:22:33 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.2.0] Succeeded in reset sshd_config.
2016/09/01 13:22:33 ERROR:[Microsoft.OSTCExtensions.VMAccessForLinux-1.4.2.0] Failed to enable the extension with error: No password or ssh_key is specified., stack trace: Traceback (most recent call last):
2016/09/01 13:22:33 ERROR:  File "/var/lib/waagent/Microsoft.OSTCExtensions.VMAccessForLinux-1.4.2.0/./vmaccess.py", line 94, in enable
2016/09/01 13:22:33 ERROR:    _set_user_account_pub_key(protect_settings, hutil)
2016/09/01 13:22:33 ERROR:  File "/var/lib/waagent/Microsoft.OSTCExtensions.VMAccessForLinux-1.4.2.0/./vmaccess.py", line 157, in _set_user_account_pub_key
2016/09/01 13:22:33 ERROR:    raise Exception("No password or ssh_key is specified.")
2016/09/01 13:22:33 ERROR:Exception: No password or ssh_key is specified.
2016/09/01 13:22:33 ERROR:
2016/09/01 13:22:33 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.2.0] Enable,error,0,Enable failed.

**Comment by boumenot** [comment]
Thank you for the report!

I don't think this is specific to your image or the version of agent.

The ordering of the calls has not changed between 1.4.1.0 (and much older) and 1.4.2.0.  The only way to get into this situation is if the configuration contains a username, but does not include a password or public key.

I need to reproduce, and capture the configuration that was pushed down by the portal.  If necessary, I will need to followup with the portal team to understand why username was included for just an SSH reset.

**Comment by boumenot** [comment]
Maybe it is specific to the image and agent.  I could not reproduce this with Ubuntu 14.04.

**Comment by boumenot** [comment]
If the image is not in the Azure Market Place I cannot reproduce this issue.  I will re-try with RHEL 7.2, and upgrade the agent to 2.1.6.

@hglkrijger - Do you have insight into this?

**Comment by hglkrijger** [comment]
@Boumenot, I did not try repro, just quickly grep'd the source. Seemed like the issue was 'things happened as expected, then we saw an error in the log which did not correlate' which made me think the call to `_set_user_account_pub_key` was unnecessary and it was expected that some of the config might not be present. If the code has not changed then perhaps it is a portal side issue?

This is the new RHEL 7.3 image which @yuxisun1217 is testing, so it won't be in the marketplace just yet.

**Comment by boumenot** [comment]
I installed RHEL 7.2, and updated the WAAgent to tip using the following bash script.  The installed agent version was 2.1.6.1.

I executed an SSH configuration reset through the portal, and I did not observe any error messages.

```sh
yum -y install git
git clone https://github.com/Azure/WALinuxAgent
cd WALinuxAgent
wget https://bootstrap.pypa.io/ez_setup.py -O - | python
python setup.py install
systemctl daemon-reload
#echo 'AutoUpdate.Enabled=n' >> /etc/waagent.conf
service waagent restart
waagent --version
tail -999f /var/log/waagent.log
```

I think the next step is to decode (decrypt) the configuration sent via the portal.  Let me know if you need instructions on how to do that.

**Comment by yuxisun1217** [comment]
Yes, I think it might be because of the different configuration sent via portal. I print the public_settings in the vm_access.py when I "reset configuration" on ARM and ASM mode to compare:
ARM mode: {u'username': u'azureuser', u'password': u'', u'ssh_key': u'', u'reset_ssh': True}
ASM mode: {u'reset_ssh': u'true'}

**Comment by boumenot** [comment]
This is odd.  I executed my test on an ARM VM, in the new (ARM) portal.  I did not observe that configuration.

Public settings should not contain those values.  The public settings are defined [here][public], and protectedsettings are defined [here][protected].

[public]: https://github.com/Azure/azure-linux-extensions/tree/master/VMAccess#11-public-configuration
[protected]: https://github.com/Azure/azure-linux-extensions/tree/master/VMAccess#12-protected-configuration

**Comment by yuxisun1217** [comment]
Hi @boumenot ,

I try to provide more details:
1. Create a RHEL-7 VM in ARM mode(I use RHEL-7.2 on-demand VM this time)
2. Install WALA-2.1.5.
3. Run "Reset Configuration Only" from the new portal:
![screenshot from 2016-09-05 15-18-04](https://cloud.githubusercontent.com/assets/19280786/18240125/6eafdad4-737c-11e6-9867-1474573406e3.png)

**Comment by boumenot** [comment]
This has been fixed in 1.4.3.0, and further refinement was made in 1.4.4.0.  Please verify.
